### PR TITLE
Dev moving window

### PR DESCRIPTION
**Pull request by Enerccio**
#40 for dev branch

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -26,13 +27,14 @@ import {
     main_api,
     online_status,
     chat_metadata,
+    setExtensionPrompt,
```
Comment (qvink) [inline index.js:30]: This is already accessible via `getContext()`

File: index.js [hunk 2]
```diff
@@ -26,13 +27,14 @@ import {
...
 } from '../../../../script.js';
 import { getContext, extension_settings, saveMetadataDebounced} from '../../../extensions.js';
 import { formatInstructModeChat, formatInstructModePrompt } from '../../../instruct-mode.js';
 import { selected_group, openGroupId } from '../../../group-chats.js';
 import { loadMovingUIState, power_user } from '../../../power-user.js';
 import { dragElement } from '../../../RossAscends-mods.js';
-import { debounce_timeout } from '../../../constants.js';
+import { debounce_timeout, inject_ids } from '../../../constants.js';
```
Comment (qvink) [inline index.js:37]: I don't see this used anywhere

File: index.js [hunk 3]
```diff
@@ -3959,12 +3972,423 @@ globalThis.memory_intercept_messages = function (chat, _contextSize, _abort, typ
...
+            }
+
+            this.short_history.pop();
+            this._generate_short_history();
+        } while (this.short_history_tokens > this.short_history_token_space);
+
+        return true;
+    }
+
+    _generate_short_history() {
```
Comment (qvink) [inline index.js:4307]: There is lots of duplicated code here, with two versions of a bunch of functions that only differ in whether you are dealing with short-term or long-term memories. These should be single functions that take in the variables to be modified.

File: index.js [hunk 4]
```diff
@@ -3959,12 +3972,423 @@ globalThis.memory_intercept_messages = function (chat, _contextSize, _abort, typ
...
+    process() {
+        let split = this.unmodified.split(SHORT_MEMORY_ROLLING_WINDOW_STARTOFHISTORY_MARKER)
+        this.pre = split[0];
+        let rest = split[1];
+        split = rest.split(SHORT_MEMORY_ROLLING_WINDOW_ENDOFHISTORY_MARKER)
+        this.messages = split[0];
+        this.messages_tokens = count_tokens(this.messages);
+        this.post = split[1];
+        this.message_buffer = this.messages;
+        this.message_buffer_tokens = count_tokens(this.messages);
```
Comment (qvink) [inline index.js:4019]: why are we counting the tokens again here?

File: index.js [hunk 5]
```diff
@@ -3959,12 +3972,423 @@ globalThis.memory_intercept_messages = function (chat, _contextSize, _abort, typ
 };
 
 
+const SHORT_MEMORY_ROLLING_WINDOW_MARKER = "@@SHORT_MEMORY_ROLLING_WINDOW_MARKER@@";
+const LONG_MEMORY_ROLLING_WINDOW_MARKER = "@@LONG_MEMORY_ROLLING_WINDOW_MARKER@@";
+const SHORT_MEMORY_ROLLING_WINDOW_STARTOFHISTORY_MARKER = "@@SHORT_MEMORY_ROLLING_WINDOW_STARTOFHISTORY_MARKER@@";
+const SHORT_MEMORY_ROLLING_WINDOW_ENDOFHISTORY_MARKER = "@@SHORT_MEMORY_ROLLING_WINDOW_ENDOFHISTORY_MARKER@@";
+
+class RollingWindowProcessor {
```
Comment (qvink) [inline index.js:3980]: It was very difficult to parse what is going on in this class, would appreciate documentation on what these variables are, what all those functions are trying to do, and the general code flow that occurs when the `process()` function is called.

File: index.js [hunk 6]
```diff
@@ -137,6 +140,8 @@ const default_settings = {
     injection_threshold_update_trigger_summaries: 0,  // update threshold when this many new summaries are ready
     injection_threshold_update_trigger_context: 0,  // update threshold when the last prompt reched this percent of max context
 
+    rolling_window: false, // whether we roll memories - in which case lot of other options below are ignored
```
Comment (qvink) [inline index.js:143]: All ignored config settings need to be disabled in the UI too, or the user won't know.

File: index.js [hunk 7]
```diff
@@ -3959,12 +3972,423 @@ globalThis.memory_intercept_messages = function (chat, _contextSize, _abort, typ
 };
 
 
+const SHORT_MEMORY_ROLLING_WINDOW_MARKER = "@@SHORT_MEMORY_ROLLING_WINDOW_MARKER@@";
```
Comment (qvink) [inline index.js:3975]: If you're using a class, might as well throw these placeholder variables in there

File: index.js [hunk 8]
```diff
@@ -3959,12 +3972,423 @@ globalThis.memory_intercept_messages = function (chat, _contextSize, _abort, typ
...
+        let filtered_post_long_history = this.post.replace(LONG_MEMORY_ROLLING_WINDOW_MARKER, long_history);
+        let filtered_post_history = filtered_post_long_history.replace(SHORT_MEMORY_ROLLING_WINDOW_MARKER, short_history);
+
+        return filtered_pre_history + this.message_buffer + filtered_post_history;
+    }
+
+
+    _find_last_message() {
+        if (this.ctx.chat.length > 0) {
+            for (let i = this.ctx.chat.length - 1; i--; i>=0) {
```
Comment (qvink) [inline index.js:4050]: This is effectively an O(m*n) loop. For each message (m) we are performing a linear search of the entire context (n) to see if it's there. It might be a minor problem with small contexts but it can definitely be optimized regardless.

File: index.js [hunk 9]
```diff
@@ -3959,12 +3972,423 @@ globalThis.memory_intercept_messages = function (chat, _contextSize, _abort, typ
...
+                this.last_message = this.last_message + 1;
+                continue;
+            }
+            let mesBufPos = this.message_buffer.indexOf(message.mes)
+            if (mesBufPos >= 0) {
+                // trim everything until mesBufPos
+                const pre = this.message_buffer.substring(0, mesBufPos);
+                this.message_buffer = this.message_buffer.substring(mesBufPos, this.message_buffer.length);
+                this.message_buffer = this.message_buffer.replace(message.mes, "")
+                this.message_buffer_tokens = count_tokens(this.message_buffer);
```
Comment (qvink) [inline index.js:4084]: This is an example of something that I would advise against:  `count_tokens()` being inside this loop worries me, especially since `this.message_buffer` is going to be changing constantly.

File: index.js [hunk 10]
```diff
@@ -3959,12 +3972,423 @@ globalThis.memory_intercept_messages = function (chat, _contextSize, _abort, typ
...
+                this.last_message = this.last_message + 1;
+
+                return count_tokens(pre + message.mes);
```
Comment (qvink) [inline index.js:4087]: Again here, prepending something to the message before counting basically guarantees a cache miss

File: index.js [hunk 11]
```diff
@@ -3959,12 +3972,423 @@ globalThis.memory_intercept_messages = function (chat, _contextSize, _abort, typ
...
+        }
+        return -1;
+    }
+
+}
+
+function update_prompt_with_rolling_memory(context, data) {
+    const processor = new RollingWindowProcessor(context, data.prompt);
+    processor.process();
+    if (processor.generated) {
```
Comment (qvink) [inline index.js:4377]: Why not just return the right output from `.process()`?

**Comment by qvink** [comment]
Thanks so much! This is a really clever method for determining which messages are in context. Let me see if I understand the process, and correct me if not:
- You inject placeholders into context
- When a new message is generated, you intercept the generation
- You search the context for the earliest message and remove it
- You generate the summary injections, repeatedly trimming it if it's too big
- As long as the summary injection is under the associated token limit, keep removing messages and adding summaries
- Finally replace the placeholders with the summary injections and continue the generation

If this is indeed the method, then I have several concerns:
1. By it's nature, this method means that summaries can no longer trigger world info because there are placeholders in context when it's searched for keywords. At the very least, the config options to include summaries in world info scans should be turned off and disabled when this feature is active (and associated warnings where appropriate).
2. Further, world info entries can be triggered by messages that are then removed, leaving world info in context that shouldn't be there. 
3. I advise against simply replacing the placeholder with empty strings if there aren't any summaries, because that leaves empty system message tags for most text completion instruct templates.
4. It looks like you assume the placeholder is always in a particular location relative to other parts of the context, but that isn't guaranteed. You can place the summaries at any depth in the chat history, which means there could be chat messages both before and after the summaries.
5. There are a *lot* of calls to `count_tokens()` when you are trying to guess at how big the summary injections need to be. Be aware that `count_tokens()` can be very a very heavy operation depending on the endpoint and I would highly advise against relying on it. If it is absolutely necessary, we want to optimize how it's used to make use of ST's token count cache whenever possible. For example, when building summary injections I estimate the total token count by adding up each summary's token count individually, never by calculating it all at once (this was a performance issue awhile ago).
6. It's not clear to me what stops the summaries from removing too many messages. It seems like the user cannot guarantee that any chat messages are kept because summaries are prioritized. One would have to guess what percent of context to have summaries take up to keep a certain amount of chat messages in context. There definitely needs to be a config setting to limit this behavior.
7. It's not clear to me how this feature prioritizes long-term vs short-term summaries when there is not enough space for both. If this is addressed, please let me know how it's handled.


Regardless of the implementation, some general items to address:
1. This PR is mixed with your lorebook macro PR, you'll need to separate these so they can be edited independently.
2. I'm confused by the name of this feature. The current way that summaries are handled is already a "rolling window", with earlier summaries getting rotated out as more are added. I think this feature requires a name that clearly delineates what it is trying to do and the tradeoffs made as a result. Most importantly, it appears to be mutually exclusive to the "Injection Threshold" method, which should be made obvious in the settings. For example, it should be clear that the user is choosing between two different methods of balancing summaries and messages. As it is, it looks like an additional option on top of everything else, even though many config settings become useless with this enabled.
3. The description given says "buffer is dynamically sized and only memories not included in the messages sent are added". First, the user has no reference for what this buffer is. Second, unless I am mistaken his doesn't seem to be true. Summaries aren't included based on messages, it's the other way around - messages are removed until you've fit a certain amount of summaries in. 
4. Make sure to test various situations and edge cases. When I tried to use the feature, my context got quite mangled and this code isn't clear enough for me to easily debug for you.
5. I've left in-line review comments.

File: index.js [hunk 12]
```diff
@@ -3990,6 +4414,11 @@ function refresh_memory() {
     ctx.setExtensionPrompt(`${MODULE_NAME}_long`,  long_injection,  long_term_position, get_settings('long_term_depth'), get_settings('long_term_scan'), get_settings('long_term_role'));
     ctx.setExtensionPrompt(`${MODULE_NAME}_short`, short_injection, short_term_position, get_settings('short_term_depth'), get_settings('short_term_scan'), get_settings('short_term_role'));
 
+    if (get_settings("rolling_window")){
+        setExtensionPrompt(`${MODULE_NAME}_rw_start`, SHORT_MEMORY_ROLLING_WINDOW_STARTOFHISTORY_MARKER, extension_prompt_types.IN_CHAT, 10000);
```
Comment (qvink) [inline index.js:4418]: use ST's exported const `MAX_INJECTION_DEPTH` instead of the hard-coded 10000

**Comment by Enerccio** [comment]
> Thanks so much! This is a really clever method for determining which messages are in context. Let me see if I understand the process, and correct me if not:
> 
>     * You inject placeholders into context
> 
>     * When a new message is generated, you intercept the generation
> 
>     * You search the context for the earliest message and remove it
> 
>     * You generate the summary injections, repeatedly trimming it if it's too big
> 
>     * As long as the summary injection is under the associated token limit, keep removing messages and adding summaries
> 
>     * Finally replace the placeholders with the summary injections and continue the generation
> 
> 
> If this is indeed the method, then I have several concerns:
> 
>     1. By it's nature, this method means that summaries can no longer trigger world info because there are placeholders in context when it's searched for keywords. At the very least, the config options to include summaries in world info scans should be turned off and disabled when this feature is active (and associated warnings where appropriate).

Yes, the option should be disabled then. 

> 
>     2. Further, world info entries can be triggered by messages that are then removed, leaving world info in context that shouldn't be there.

That is an issue but I am unsure if it can be fixed. 

> 
>     3. I advise against simply replacing the placeholder with empty strings if there aren't any summaries, because that leaves empty system message tags for most text completion instruct templates.

Is there anything else that can be done? I could try removing the system tag around it as well if I can get it from template.

>     4. It looks like you assume the placeholder is always in a particular location relative to other parts of the context, but that isn't guaranteed. You can place the summaries at any depth in the chat history, which means there could be chat messages both before and after the summaries.

Yes, if you want to use this option then insertion point must not be in the chat history itself (who would do that option though? That makes zero sense...)

> 
>     5. There are a _lot_ of calls to `count_tokens()` when you are trying to guess at how big the summary injections need to be. Be aware that `count_tokens()` can be very a very heavy operation depending on the endpoint and I would highly advise against relying on it. If it is absolutely necessary, we want to optimize how it's used to make use of ST's token count cache whenever possible. For example, when building summary injections I estimate the total token count by adding up each summary's token count individually, never by calculating it all at once (this was a performance issue awhile ago).

I can see if I can reduce number of count_token calls but with 32k context even if my service is in the USA and I am in Europe I see no noticeable delay.

> 
>     6. It's not clear to me what stops the summaries from removing too many messages. It seems like the user cannot guarantee that any chat messages are kept because summaries are prioritized. One would have to guess what percent of context to have summaries take up to keep a certain amount of chat messages in context. There definitely needs to be a config setting to limit this behavior.

Summaries only remove up to the token limit you set for each summary type. They are removed FIFO so it basically emulates what would happen if you filled summaries first. 

> 
>     7. It's not clear to me how this feature prioritizes long-term vs short-term summaries when there is not enough space for both. If this is addressed, please let me know how it's handled.

It inserts short term memories first until the buffer for them is full. After that, it moves to the long term memories, shifting short term memories up. Here is example of how it works with message ids

Summaries [90L, 91L, 92, 93L, 94, 95, 96, 97L, 98, 99, 100, 101, 102, 103, 104]
Messages [100, 101, 102, 103, 104, ...]
Short term []
Long term []

remove message 100 to free space for short term summaries

Messages [101, 102, 103, 104, ...]
Short term [98, 99, 100]
Long term []

Remove 101 for more space

Messages [102, 103, 104, ...]
Short term [98, 99, 100, 101]
Long term []


Messages [102, 103, 104, ...]
Short term [96, 97, 98, 99, 100, 101]
Long term []

lets say we hit the limit, so move to long term, free space for long term (possibly remove short term if it overflows short term)

Messages [103, 104, ...]
Short term [97, 98, 99, 100, 101, 102]
Long term []

find first long term memory upstream from 97

Messages [103, 104, ...]
Short term [97, 98, 99, 100, 101, 102]
Long term [93L]

We need more space for long term, inserted 103 at the start of short term, removed 97, 98 to keep size of short term

Messages [104, ...]
Short term [99, 100, 101, 102, 103]
Long term [93L]

Find nearest long term from 99, add it to the start

Messages [104, ...]
Short term [98, 99, 100, 101, 102, 103]
Long term [93L, 98L]

Find more long terms

Messages [104, ...]
Short term [98, 99, 100, 101, 102, 103]
Long term [91L, 93L, 98L]

until both buffers are full, then concatenate and insert into prompt.
> 
> 
> Regardless of the implementation, some general items to address:
> 
>     1. This PR is mixed with your lorebook macro PR, you'll need to separate these so they can be edited independently.

Done with other branch now

> 
>     2. I'm confused by the name of this feature. The current way that summaries are handled is already a "rolling window", with earlier summaries getting rotated out as more are added. I think this feature requires a name that clearly delineates what it is trying to do and the tradeoffs made as a result. Most importantly, it appears to be mutually exclusive to the "Injection Threshold" method, which should be made obvious in the settings. For example, it should be clear that the user is choosing between two different methods of balancing summaries and messages. As it is, it looks like an additional option on top of everything else, even though many config settings become useless with this enabled.

I could rename it to DynamicMemory or something like that? Yes, it should disable more options in the html settings. Honestly though, after using it for few days I think this is preferred way to do it despite the tradeoffs (which honestly is only the worldinfo triggering messages disappearing) - I use maximum actual messages and only then short/long term is inserted.

> 
>     3. The description given says "buffer is dynamically sized and only memories not included in the messages sent are added". First, the user has no reference for what this buffer is. Second, unless I am mistaken his doesn't seem to be true. Summaries aren't included based on messages, it's the other way around - messages are removed until you've fit a certain amount of summaries in.
> 

Maybe I described it wrong. Dynamically sized - if you don't have enough memories, you don't waste short/long term token size and instead more of real messages get it. The second part of my sentence is that only summaries for the messages not included are sent unlike how it is now where you get your message and it's summary sent together in same query. 

>     4. Make sure to test various situations and edge cases. When I tried to use the feature, my context got quite mangled and this code isn't clear enough for me to easily debug for you.

I could try but unless you mess up with insertion positions it should work. 
> 
>     5. I've left in-line review comments.

Will check them.

File: index.js (hunk 8 above)
Comment (Enerccio) [reply index.js:4050]: how would you optimize it? We need to find first message that was included in the prompt

File: index.js (hunk 11 above)
Comment (Enerccio) [reply index.js:4377]: Because process can end on multiple places, it would have to be another method anyways?

File: index.js (hunk 10 above)
Comment (Enerccio) [reply index.js:4087]: changed it to simple subtraction

File: index.js (hunk 9 above)
Comment (Enerccio) [reply index.js:4084]: Well we have to know how many tokens we removed

File: index.js (hunk 7 above)
Comment (Enerccio) [reply index.js:3975]: Done

File: index.js (hunk 4 above)
Comment (Enerccio) [reply index.js:4019]: Fixed

File: index.js (hunk 5 above)
Comment (Enerccio) [reply index.js:3980]: Done

**Comment by Enerccio** [comment]
Moved it to new branch #47

File: index.js (hunk 8 above)
Comment (qvink) [reply index.js:4050]: Take the text from the message buffer that you know to be from the first message and compare it to the individual messages. That way you are at least not comparing to the entire message buffer each time. Then you could do a binary search from there.

File: index.js (hunk 9 above)
Comment (qvink) [reply index.js:4084]: Where is `this.message_buffer_tokens` used?

File: index.js (hunk 11 above)
Comment (qvink) [reply index.js:4377]: Just use a flag. Since this conditional check is required after `process()` is called it should just be a part of the function.

**Comment by qvink** [comment]
> Yes, the option should be disabled then.

Not only that, but you would need to disable all the extra computation that is done for those options.

> > 2. Further, world info entries can be triggered by messages that are then removed, leaving world info in context that shouldn't be there.

> That is an issue but I am unsure if it can be fixed.

I agree, this is a fundamental limitation of the feature. Combined with summaries no longer triggering world info is (to me) a huge issue. This feature might allow you to optimize your context, but this seriously messes with context in other ways, potentially offsetting the benefits.

>> 3. I advise against simply replacing the placeholder with empty strings if there aren't any summaries, because that leaves empty system message tags for most text completion instruct templates.

> Is there anything else that can be done? I could try removing the system tag around it as well if I can get it from template.

I think that's what you'll have to do. There isn't a built-in way to accomplish this because you are already doing something very unintended by editing injections after they are generated. You will need to make sure it's a generalized solution, because it's not just system tags, it could be anything.

>> 4. It looks like you assume the placeholder is always in a particular location relative to other parts of the context, but that isn't guaranteed. You can place the summaries at any depth in the chat history, which means there could be chat messages both before and after the summaries.

> Yes, if you want to use this option then insertion point must not be in the chat history itself (who would do that option though? That makes zero sense...)

Just because it doesn't make sense to you doesn't mean you can ignore it. Not only that, but users can insert the summaries *anywhere* using macros, which this also doesn't account for.

> I can see if I can reduce number of count_token calls but with 32k context even if my service is in the USA and I am in Europe I see no noticeable delay.

It's not about delay, it's that counting tokens is an inherently heavy operation. The problem is that you perform a linear search of how many summaries to inject, re-checking the count each time. Instead, you need to calculate the size of individual summaries ahead of time to make use of the token count cache.


>> 6. It's not clear to me what stops the summaries from removing too many messages. It seems like the user cannot guarantee that any chat messages are kept because summaries are prioritized. One would have to guess what percent of context to have summaries take up to keep a certain amount of chat messages in context. There definitely needs to be a config setting to limit this behavior.

> Summaries only remove up to the token limit you set for each summary type. They are removed FIFO so it basically emulates what would happen if you filled summaries first.

Yes, but if your context is almost full this might remove all your messages. You need a secondary limit for what the minimum number of messages to keep is. For example, I might want to keep at least 5 messages in context, even if that means not having all the summaries. 

> It inserts short term memories first until the buffer for them is full. After that, it moves to the long term memories, shifting short term memories up. Here is example of how it works with message ids

Great, makes sense

> I could rename it to DynamicMemory or something like that? 

> Maybe I described it wrong. Dynamically sized - if you don't have enough memories, you don't waste short/long term token size and instead more of real messages get it.

The current implementation of memory is already dynamically sized based on the number of summaries you have. The thing that's dynamic about this feature is the *number of messages* injected because that changes depending on how many summaries there are. This is in contrast to the hard cutoff of the injection threshold that is currently used.

> The second part of my sentence is that only summaries for the messages not included are sent unlike how it is now where you get your message and it's summary sent together in same query.

But again, that's not what happens. It's not "only summaries for messages not included are sent", it's "only **messages** after the most recent summary are sent". And right now, messages and summaries **aren't** sent together if you use the injection threshold with `Remove Messages`.


Ok let's start from the beginning. As I understand it, your original problem is this: "When we use the `Injection Threshold` with `Remove Messages`, then there is a hard cutoff to the number of messages sent, which can lead to unused context". (again, summaries are **NOT** sent with their associated message). First, the point of this extension isn't to completely fill up your context, it's to *reduce the amount of information in context*. LLMs (especially smaller models) generally perform better with less information and an under-full context. When using the `Injection Threshold`, the entire point is to limit how much of your chat is in context and keep only the important information via summaries. This reduces the tokens used and makes information easier to retrieve for the LLM. With your new feature, all you are doing is *sometimes* maximizing the number of messages in context, which is not even necessarily beneficial. And this is done at the cost of world info, text template integrity, summary placement, context update triggers, etc. Half the features of the extension become useless just so you can *sometimes* get more messages, which I emphasize is *not always a good thing*. To me this just doesn't seem worth it for most users, and it's just a huge change for very niche use and limited benefit.
